### PR TITLE
feat: add Melon Protocol

### PR DESCRIPTION
**Pull request by developerfred**
Add Melon Protocol

[Docs](https://docs.melonport.com/)

### CODE REVIEW

File: migrations_scripts/1_deploy_registry_and_add_adapters.js [hunk 1]
```diff
@@ -151,6 +153,9 @@ const iETHAddress = '0xA9859874e1743A32409f75bB11549892138BBA1E';
 
 const wethAddress = '0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2';
 
+const melonAddress = '0xec67005c4e498ec7f55e092bd1d35cbc47c91892';
+const mlnfAddress = '0xd485c308df50f13cb1ec06d5fd3a44a0ae49c7fc';
```
Comment (fubhy) [inline migrations_scripts/1_deploy_registry_and_add_adapters.js:157]: This is the MLNF (Melon Fund) token of a specific fund, not globally for the network. The MLNF token is the Shares contract of a fund and one instance of this contract is deployed per fund. E.g. this one belongs to the fund with the address `0x37d5c61240d163d1684cd66e052a81be596dc6ff`.

Comment (developerfred) [reply migrations_scripts/1_deploy_registry_and_add_adapters.js:157]: Thanks @fubhy, I will change this.

Comment (developerfred) [reply migrations_scripts/1_deploy_registry_and_add_adapters.js]: fixed it here 4ead7a6

File: contracts/adapters/melonProtocol/MelonTokenAdapter.sol [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+
+/**
+ * @dev Accounting contract interface.
+ * Only the functions required for MelonAdapter contract are added.
+ * The Accounting contract is available here
+ * github.com/melonproject/protocol/blob/master/src/fund/accounting/Accounting.sol.
+ */
+interface Accounting{
+    function getFundHoldings() external view returns (uint[] memory, address[] memory);
+    function calcAssetGAV(address) external view returns (uint);
```
Comment (sobolev-igor) [inline contracts/adapters/melonProtocol/MelonTokenAdapter.sol:55]:
```suggestion
```

Comment (sobolev-igor) [reply contracts/adapters/melonProtocol/MelonTokenAdapter.sol:55]: not used

File: contracts/adapters/melonProtocol/MelonTokenAdapter.sol [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+    }
+
+    /**
+     * @return Array of Component structs with underlying tokens rates for the given token.
+     * @dev Implementation of TokenAdapter interface function.
+     */
+    function getComponents(address token) external view override returns (Component[] memory) {  
+        uint[] memory quantities;
+        address[] memory fundHoldings;
+        try Accounting(token).getFundHoldings() returns (uint[] memory balance, address[] memory result) {
```
Comment (sobolev-igor) [inline contracts/adapters/melonProtocol/MelonTokenAdapter.sol:87]: `token` is not accounting address
accounting address should be retrieved with the hub contract

File: contracts/adapters/melonProtocol/MelonTokenAdapter.sol [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+        } catch {
+            fundHoldings = new address[](0);
+            quantities = 0;
+        }
+
+        uint256 totalSupply = ERC20(token).totalSupply();
+
+        Component[] memory underlyingTokens = new Component[](fundHoldings.length);
+
+        address underlyingToken;
```
Comment (sobolev-igor) [inline contracts/adapters/melonProtocol/MelonTokenAdapter.sol:99]: there is no need to use `underlyingToken` variable

File: contracts/adapters/melonProtocol/MelonTokenAdapter.sol [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+        for (uint256 i = 0; i < fundHoldings.length; i++) {
+            underlyingToken = fundHoldings[i];
+
+            underlyingTokens[i] = Component({
+                token: underlyingToken,
```
Comment (sobolev-igor) [inline contracts/adapters/melonProtocol/MelonTokenAdapter.sol:104]:
```suggestion
                token: fundHoldings[i],
```

File: contracts/adapters/melonProtocol/MelonTokenAdapter.sol [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+    address internal constant REGISTRY = 0xb9Cb55C9366a224647B7ff66252b3613185DA0B9;
+
+    /**
+     * @return TokenMetadata struct with ERC20-style token info.
+     * @dev Implementation of TokenAdapter interface function.
+     */
+    function getMetadata(address token) external view override returns (TokenMetadata memory) {
+        return TokenMetadata({
+            token: token,
+            name: Registry(REGISTRY).getName(token),
```
Comment (sobolev-igor) [inline contracts/adapters/melonProtocol/MelonTokenAdapter.sol:74]: This is wrong

name should be taken from `name()` function of MLNF token, but is should be chenged into readable foramat, i.e. 
name for `0x3F82F3Ba33f50124eC27b2fB813230282d432971` is `0x000000000000000000000000000000000000000000004d697261692046756e64`
it should be changed into `'Mirai Fund'`

Comment (sobolev-igor) [reply contracts/adapters/melonProtocol/MelonTokenAdapter.sol:74]:
```
    function convert(string memory inputString) internal pure returns (string memory) {
        bytes memory inputBytes = bytes(inputString);
        uint256 resultLength = 0;
        
        for (uint256 i = 0; i < inputBytes.length; i++) {
            if (inputBytes[i] != bytes1(0)) {
                resultLength++;
            }
        }
        
        bytes memory result = new bytes(resultLength);
        uint256 counter = 0;
        
        for (uint256 i = 0; i < inputBytes.length; i++) {
            if (inputBytes[i] != bytes1(0)) {
                result[counter] = inputBytes[i];
                counter++;
            }
        }
        
        return string(result);
    }
```

File: contracts/adapters/melonProtocol/MelonTokenAdapter.sol (hunk 4 above)
Comment (developerfred) [reply contracts/adapters/melonProtocol/MelonTokenAdapter.sol:99]: done

File: contracts/adapters/melonProtocol/MelonTokenAdapter.sol (hunk 5 above)
Comment (developerfred) [reply contracts/adapters/melonProtocol/MelonTokenAdapter.sol:104]: done [x]

File: contracts/adapters/melonProtocol/MelonTokenAdapter.sol (hunk 2 above)
Comment (developerfred) [reply contracts/adapters/melonProtocol/MelonTokenAdapter.sol:55]: done

File: contracts/adapters/melonProtocol/MelonTokenAdapter.sol (hunk 3 above)
Comment (developerfred) [reply contracts/adapters/melonProtocol/MelonTokenAdapter.sol:87]: done

File: migrations_scripts/1_deploy_registry_and_add_adapters.js [hunk 7]
```diff
@@ -23,7 +23,12 @@ const DSRAdapter = artifacts.require('DSRAdapter');
 const GovernanceAdapter = artifacts.require('GovernanceAdapter');
 const MCDAssetAdapter = artifacts.require('MCDAssetAdapter');
 const MCDDebtAdapter = artifacts.require('MCDDebtAdapter');
+<<<<<<< HEAD
```
Comment (fubhy) [inline migrations_scripts/1_deploy_registry_and_add_adapters.js:26]: Merge conflict leftovers.

Comment (developerfred) [reply migrations_scripts/1_deploy_registry_and_add_adapters.js]: Done on Merge a5e1d98

**Pull request merged**